html_xblock/html.py

```diff
@@ -177,22 +177,55 @@ def get_editor_plugins(self):
             plugin: self.runtime.local_resource_url(self, plugin_path.format(plugin=plugin)) for plugin in plugins
         }
 
+    def substitute_keywords(self):
+        """
+        Replaces all %%-encoded words using KEYWORD_FUNCTION_MAP mapping functions.
+
+        Iterates through all keywords that must be substituted and replaces them by calling the corresponding functions
+        stored in `keywords`. If the function throws a specified exception, the substitution is not performed.
+
+        Functions stored in `keywords` must either:
+            - return a replacement string
+            - throw `KeyError` or `AttributeError`, `TypeError`.
+        """
+        data = self.data
+        system = getattr(self, 'system', None)
+        if not system:  # This shouldn't happen, but if `system` is missing, then skip substituting keywords.
+            return data
+
+        keywords = {
+            '%%USER_ID%%': lambda: getattr(system, 'anonymous_student_id'),
+            '%%COURSE_ID%%': lambda: getattr(system, 'course_id').html_id(),
```

Comment:
Do we expect these to change during the invocation? Not sure why the `lambda` is necessary.

Reply:
`lambda` approach is more flexible when it comes to extending this and ensures that all values are lazily-evaluated, so if we added something that takes time to calculate, we wouldn't waste time on doing it, if not needed.

Reply:
I see.